From d795dda495a204a0603a2a99b7ef9a158fb2df77 Mon Sep 17 00:00:00 2001
From: Zahari Doychev <zahari.doychev@intel.com>
Date: Fri, 22 May 2020 14:10:35 +0200
Subject: [PATCH] pon_qos: delay port threshold configuration

Configure the port thresholds when the first scheduler is allocated otherwise
DP returns error because the port is not configured.

Signed-off-by: Zahari Doychev <zahari.doychev@intel.com>
---
 drivers/net/datapath/pon_qos/pon_qos_tc_qos.c | 19 ++++++++++---------
 1 file changed, 10 insertions(+), 9 deletions(-)

diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_qos.c b/drivers/net/datapath/pon_qos/pon_qos_tc_qos.c
index c1ff8079b59a..f79cbd50ede1 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_qos.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_qos.c
@@ -58,8 +58,9 @@ static int pon_qos_overwrite_port_thresholds(struct pon_qos_qdisc *sch)
 		.green_threshold = PORT_GREEN_TH,
 		.yellow_threshold = PORT_YELLOW_TH,
 	};
+	int flags = DP_PORT_CFG_GREEN_THRESHOLD | DP_PORT_CFG_YELLOW_THRESHOLD;
 
-	if (dp_qos_port_conf_set(&port_cfg_info, 0) == DP_FAILURE) {
+	if (dp_qos_port_conf_set(&port_cfg_info, flags) == DP_FAILURE) {
 		netdev_err(sch->dev, "%s: dp_qos_port_conf_set() failed\n",
 			   __func__);
 		return -1;
@@ -98,14 +99,6 @@ int pon_qos_fill_port_data(struct pon_qos_qdisc *sch)
 	sch->def_q = q_res[0].q_id;
 	sch->epn = deq.cqm_deq_port;
 
-	if (sch->deq_idx >= 0) {
-		if (pon_qos_overwrite_port_thresholds(sch)) {
-			netdev_err(sch->dev, "%s: pon_qos_overwrite_port_thresholds() failed\n",
-				   __func__);
-			return -3;
-		}
-	}
-
 	return 0;
 }
 
@@ -251,6 +244,14 @@ int pon_qos_add_sched(struct pon_qos_qdisc *sch, int prio)
 		return -1;
 	}
 
+	if (sch->deq_idx >= 0) {
+		if (pon_qos_overwrite_port_thresholds(sch)) {
+			netdev_err(sch->dev, "%s: pon_qos_overwrite_port_thresholds() failed\n",
+				   __func__);
+			return -3;
+		}
+	}
+
 	return 0;
 }
 
